lib/oplog/last_minute_metric.go

```diff
@@ -0,0 +1,265 @@
+	value        float64
+	bucketedTime time.Time
+}
+
+// DefaultInterval is the default interval for the
+const DefaultInterval = 1 * time.Minute
+
+// IntervalMaxOpts are options for IntervalMaxMetric.
+type IntervalMaxOpts struct {
+	prometheus.Opts
+
+	// ReportInterval is the interval by which reports will be bucketed. Default 1m.
+	ReportInterval time.Duration
+}
+
+// NewIntervalMaxMetric constructs a new IntervalMaxMetric.
+func NewIntervalMaxMetric(opts IntervalMaxOpts, labels []string, labelValues []string) *IntervalMaxMetric {
+	if opts.ReportInterval == 0 {
+		opts.ReportInterval = DefaultInterval
+	}
+
+	return &IntervalMaxMetric{
+		desc: prometheus.NewDesc(
+			prometheus.BuildFQName(
+				opts.Namespace, opts.Subsystem, opts.Name,
+			),
+			opts.Help,
+			labels,
+			opts.ConstLabels,
+		),
+		opts:        opts,
+		labelValues: labelValues,
+
+		currentMax: nil,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *IntervalMaxMetric) Describe(descs chan<- *prometheus.Desc) {
+	descs <- c.desc
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *IntervalMaxMetric) Collect(mtcs chan<- prometheus.Metric) {
+	c.lck.Lock()
+	defer c.lck.Unlock()
+
+	currentBucket := c.opts.thisTimeBucket()
+	c.rotate(currentBucket)
+
+	if c.previousMax == nil {
+		return
+	}
+
+	if currentBucket.Sub(c.previousMax.bucketedTime) != c.opts.ReportInterval {
+		return
+	}
+
+	mtcs <- prometheus.MustNewConstMetric(c.desc, prometheus.GaugeValue, c.previousMax.value, c.labelValues...)
+}
+
+// Report reports a value to the IntervalMaxMetric. If it is the greatest seen so far in the current window, it will be
+// recorded as such until either another, greater value is reported, or the window ends (it will then be the
+// value this IntervalMaxMetric returns when polled via Collect, until another window elapses).
+func (c *IntervalMaxMetric) Report(value float64) {
+	c.lck.Lock()
+	defer c.lck.Unlock()
+
+	thisTimeBucket := c.opts.thisTimeBucket()
+	c.rotate(thisTimeBucket)
+
+	maxVal := &lastMax{
+		value:        value,
+		bucketedTime: thisTimeBucket,
+	}
+
+	if c.currentMax == nil {
+		c.currentMax = maxVal
+		return
+	}
+
+	if thisTimeBucket.Equal(c.currentMax.bucketedTime) {
+		if c.currentMax.value < value {
+			c.currentMax = maxVal
+		}
+
+		return
+	}
+
+	if thisTimeBucket.After(c.currentMax.bucketedTime) {
+		c.currentMax = maxVal
+		return
+	}
+
+	// this bucket is before previous bucket
+	panic("interval max metric time traveled")
+}
+
+// pre: c is locked
+func (c *IntervalMaxMetric) rotate(timeBucket time.Time) {
+	if c.currentMax == nil {
+		return
+	}
+
+	if !timeBucket.After(c.currentMax.bucketedTime) {
+		return
+	}
+
+	// this behavior is expected by IntervalMaxMetricVec: must set currentMax nil if we've advanced a bucket
+	c.previousMax = c.currentMax
+	c.currentMax = nil
+}
+
+func (opts IntervalMaxOpts) thisTimeBucket() time.Time {
+	return time.Now().Truncate(opts.ReportInterval)
```

Comment:
I think this might cause problems if the local clock changes (maybe even running to the "time traveling" panic above).

We might be able to solve that by using the monotonic clock -- instead of rounding based on actual time, we could do `opts.startTime := time.Now()` in the initializer, then then in here do `(time.Since(opts.startTime)).Seconds() / opts.ReportInterval.Seconds()` -- my understanding of the Go time system is that the monotonic clock is used when subtracting times from each other. This means we lose absolute minute boundaries (which would perhaps cause confusing behavior when the service restarts?) but we shouldn't be impacted by a local clock change

Reply:
I took this approach, but with syncing to the wall-clock interval boundary in the constructor. This will still drift over time, but it shouldn't be too much. At worst we get a little bit of discontinuity when the service restarts, or conceivably a rare double-report if our scrape interval matches up really precisely with the interval's.